### ISSUE TITLE
exclusive cart rules

### ISSUE THREAD

**Issue by centoasa** [PrestaShop/PrestaShop#13243]
Prestashop 1.7.5.1 on debian
The current cart rules state that there is at least one product of those selected in the rule, ie if I have in the cart 1 product of the rule and others that have no rules, the rule is applied.
Is there not the possibility of having the rule exclusively for the selected product? That is, if I select a certain brand, product or category for the rule, I would like the rule to be applied only if the cart contains only 1 or more products of that specific product or brand. I hope I'm clear.

### Steps to reproduce
1. Go to BO > Catalog > Discounts > Add new cart rule
2. In Information > Add a name
3. In Conditions > Check Product selection and Choose any product
4. In Actions > Choose Percent (%) and leave it empty + Apply a discount to Selected product(s)
5. Click Save and stay > Successful creation
6. Go to Tab Actions
7. See Apply a discount is checked to "None" :x:

**Comment by khouloudbelguith** [comment]
Hi @centoasa,

I did not manage to reproduce the issue with PS:1.7.5.1.
I attached a video record.
https://drive.google.com/file/d/1KwLjrR2-ubpB4tbhq3J2Ba5uAdVRzIMZ/view

Thanks to check and feedback.

**Comment by centoasa** [comment]
In my case, I would a cart rules only to have free shipping only for a single brand. 
If I select "free shipping" button, and select one of "apply discount" also for 0%, the rule will not saved.
https://nimb.ws/t6yJA7
see also my video to explain better: https://nimb.ws/dSREfQ

**Comment by khouloudbelguith** [comment]
@centoasa, the percentage value must be greater than 0 otherwise no discount.

**Comment by centoasa** [comment]
this is right @khouloudbelguith ! But how can I have cart rule only for free shipping, only for a single brand or product?

**Comment by khouloudbelguith** [comment]
@centoasa, I attached a video record.
https://drive.google.com/file/d/1Nqv1zLS_wBU8FD0KwBG9s2UQ8RfsNhCB/view

Thanks to check and feedback.

**Comment by centoasa** [comment]
@khouloudbelguith 
Try this scenario: courier not free, but for a fee (you choose the rate); product that can be shipped only through this paid carrier; shopping cart rule for this product; then add to the cart this product and another randomly chosen one, obviously that does not belong to any cart rule. You will see, however, that the cart will be free, because the current "cart rule" issue is: "at least one product / brand / category" that is contained in the cart makes the rule apply. Instead, in my opinion, an admin should choose this option, or "rule cart only and exclusively for this product / brand / category (if you put another product the rule is nullified). I hope I was clear.

**Comment by khouloudbelguith** [comment]
@centoasa, I manage to reproduce the issue with PS:1.7.5.1
I attached a video record.
https://drive.google.com/file/d/1_mDZYQuLeXCdrl6gDemEzqHlYp49NB6y/view
This issue is added to our bug roadmap.

Thanks!

**Comment by khouloudbelguith** [comment]
Same issue with PS1770beta2build1
![image](https://user-images.githubusercontent.com/16067358/91168055-40ee4100-e6d5-11ea-92e4-240940ec4661.png)
=> Save
![image](https://user-images.githubusercontent.com/16067358/91168086-52374d80-e6d5-11ea-8fae-84e9af6233ec.png)

Thanks!

**Comment by boo-code** [comment]
Reading the current code, the behaviour @khouloudbelguith reproduced follows from the design rather than from a defect, and that matters for what a fix would look like.

`product_restriction` is a Conditions field. `CartRule::checkValidity()` uses it as a gate: `checkProductRestrictionsFromCart()` decides whether the rule applies at all. Once it passes, `getContextualValue()` grants the shipping discount over the whole cart:

```php
if ($this->free_shipping && in_array($filter, [..., CartRule::FILTER_ACTION_SHIPPING])) {
    if (!$this->carrier_restriction) {
        $reduction_carrier += $context->cart->getOrderTotal($use_tax, Cart::ONLY_SHIPPING, ...);
```

There is no reference there to which products matched, and nothing in between narrows the shipping to them. So "at least one product from the selection" is the intended reading of the condition, and the request here, apply only when the cart contains nothing else, is a new option rather than a correction of a broken one.

That distinction decides where it should be built. The discount domain is mid-migration: `install-dev/data/xml/feature_flag.xml` declares `discount` with `state="0"` and `stability="beta"`, described as "the new discount system", and `CartRuleCalculator` already branches on it. The new model has explicit types in `DiscountType` - `product_level`, `cart_level`, `free_shipping`, `free_gift` - with a priority order between them, and it keeps free shipping as a shipping-level discount rather than a per-product one, so it does not give this for free either.

Adding an exclusivity flag to `ps_cart_rule` now means a new legacy column, a BO form field, validation and an upgrade script on the model that is being replaced. I would rather not do that unasked. If the option is wanted, is the new discount system the right place for it, and should it apply to every discount type or only to free shipping? With that settled I will implement it there.
